### PR TITLE
refactor: helper functions to generate placement structs

### PR DESCRIPTION
**Pull request by rajatsing**
- Added helper functions to generate placement structs and added more function params to change values for a specific struct. 

- This will make each struct more flexible and will also reduce the complexity of structures that need to be written by hand.

Signed-off-by: RAJAT SINGH <rajasing@redhat.com>

### CODE REVIEW

File: pkg/controller/defaults/placements.go [hunk 1]
```diff
@@ -248,3 +91,70 @@ func getTopologySpreadConstraintsSpec(selectorValue string) []corev1.TopologySpr
 
 	return topologySpreadConstraints
 }
+
+func setWeightedPodAffinityTerm(selector string) []corev1.WeightedPodAffinityTerm {
```
Comment (jarrpa) [inline pkg/controller/defaults/placements.go:95]:
```suggestion
func getWeightedPodAffinityTerm(selector string) []corev1.WeightedPodAffinityTerm {
```

File: pkg/controller/defaults/placements.go [hunk 2]
```diff
@@ -248,3 +91,70 @@ func getTopologySpreadConstraintsSpec(selectorValue string) []corev1.TopologySpr
...
+					},
+				},
+				TopologyKey: corev1.LabelHostname,
+			},
+		},
+	}
+	return WeightedPodAffinityTerm
+}
+
+func setPodAffinityTerm(selector string) []corev1.PodAffinityTerm {
```
Comment (jarrpa) [inline pkg/controller/defaults/placements.go:116]:
```suggestion
func getPodAffinityTerm(selector string) []corev1.PodAffinityTerm {
```

File: pkg/controller/defaults/placements.go [hunk 3]
```diff
@@ -248,3 +91,70 @@ func getTopologySpreadConstraintsSpec(selectorValue string) []corev1.TopologySpr
...
+					},
+				},
+			},
+			TopologyKey: corev1.LabelHostname,
+		},
+	}
+	return podAffinityTerm
+}
+
+func setToleration() []corev1.Toleration {
```
Comment (jarrpa) [inline pkg/controller/defaults/placements.go:134]:
```suggestion
func getOcsToleration() []corev1.Toleration {
```

File: pkg/controller/defaults/placements.go [hunk 4]
```diff
@@ -248,3 +91,70 @@ func getTopologySpreadConstraintsSpec(selectorValue string) []corev1.TopologySpr
...
+			Key:      NodeTolerationKey,
+			Operator: corev1.TolerationOpEqual,
+			Value:    "true",
+			Effect:   corev1.TaintEffectNoSchedule,
+		},
+	}
+	return toleration
+}
+
+func setNodeSelector() *corev1.NodeSelector {
```
Comment (jarrpa) [inline pkg/controller/defaults/placements.go:146]:
```suggestion
func getOcsNodeSelector() *corev1.NodeSelector {
```

**Review by jarrpa — CHANGES_REQUESTED** [review]
Looks handy! The function names could use some work. Also, try and avoid generic commit descriptions. Try something like "refactor: helper functions to generate placement structs"

File: pkg/controller/defaults/placements.go (hunk 1 above)
Comment (rajatsing) [reply pkg/controller/defaults/placements.go]: Done!

File: pkg/controller/defaults/placements.go (hunk 2 above)
Comment (rajatsing) [reply pkg/controller/defaults/placements.go]: Done

File: pkg/controller/defaults/placements.go (hunk 3 above)
Comment (rajatsing) [reply pkg/controller/defaults/placements.go]: Done

File: pkg/controller/defaults/placements.go (hunk 4 above)
Comment (rajatsing) [reply pkg/controller/defaults/placements.go]: Done

**Review by umangachapagain — CHANGES_REQUESTED** [review]
I don't see any value add to this.

I prefer duplicate code (for something as static as this) rather than a bunch of function calls.
We could just define global vars/constants and re-use it (if duplication is a concern)? Saves a lot of calls.

Unless I'm completely missing the point, I don't agree with this refactor as it is.

**Comment by rajatsing** [comment]
@umangachapagain  above is [disscussed here](https://github.com/openshift/ocs-operator/pull/727/files#r485077430) and [here](https://github.com/openshift/ocs-operator/pull/727/files#r485076961)

**Comment by jarrpa** [comment]
Why is reducing duplication not inherently valuable? I believe we've already run into scenarios where we forgot to update some components when we decided to implement a change that was supposed to affect all components.

The functions also allow us to generate structs that are largely identical save for one field. If  we did it with vars/consts we'd have large structs that were largely the same. Again, why is this not valuable?

**Comment by rajatsing** [comment]
@umangachapagain do you have any reviews in it , otherwise @jarrpa  can go ahead and merge this!

**Comment by umangachapagain** [comment]
> Why is reducing duplication not inherently valuable?

It's never inherently valuable. Even if it is, it's negligible in this case.

> I believe we've already run into scenarios where we forgot to update some components when we decided to implement a change that was supposed to affect all components.

I don't see this re-factor helping us fix those issues. Even with this re-factor one can mess it up (vs the variables which was pretty obvious).

What about when we want to make changes exclusive to a single component? We then add a bunch of if..else which will then blow out of proportion? 

> The functions also allow us to generate structs that are largely identical save for one field. If we did it with vars/consts we'd have large structs that were largely the same.

But, is it always going to be identical? Is there no possibility that we'd maybe like to configure placement for each component differently (other than the name)?

> Again, why is this not valuable?

It's not "not valuable". It's just not significantly valuable. Here's why I think so

+ it doesn't significantly simplify the code
+ it doesn't make it easier to custom configure each component (we get what the function throws at us)
+ retrieving values would require less cycles then generating values
+ this file is still largely static. The configuration would probably change once in a release or two (functions just feels like an overkill)

I'm not against this re-factor, but I believe it's half way there.
It reduced duplication at the cost of configurability and extra processing cycle.
If it can give me configurability, I'd be more than happy with it. But, that's just me.

Since it's not a very big change, I'm not blocking this with my comment.
I'd like the value add of this refactor to be explained (PR and Commit descriptions) and justified rather than just calling it a "refactor".

**Comment by umangachapagain** [comment]
> @umangachapagain above is [disscussed here](https://github.com/openshift/ocs-operator/pull/727/files#r485077430) and [here](https://github.com/openshift/ocs-operator/pull/727/files#r485076961)

Can't reach those comments.

File: pkg/controller/defaults/placements.go [hunk 5]
```diff
@@ -11,219 +11,62 @@ var (
...
-										Operator: metav1.LabelSelectorOpIn,
-										Values:   []string{"rook-ceph-mon"},
-									},
-								},
-							},
-							TopologyKey: corev1.LabelHostname,
-						},
-					},
-				},
+				PreferredDuringSchedulingIgnoredDuringExecution: getWeightedPodAffinityTerm("rook-ceph-mon"),
```
Comment (jarrpa) [inline pkg/controller/defaults/placements.go:25]:
```suggestion
				PreferredDuringSchedulingIgnoredDuringExecution: []corev1.WeightedPodAffinityTerm{
					getAppWeightedPodAffinityTerm(100, "rook-ceph-mon"),
				}
```

File: pkg/controller/defaults/placements.go [hunk 6]
```diff
@@ -248,3 +91,70 @@ func getTopologySpreadConstraintsSpec(selectorValue string) []corev1.TopologySpr
 
 	return topologySpreadConstraints
 }
+
+func getWeightedPodAffinityTerm(selector string) []corev1.WeightedPodAffinityTerm {
```
Comment (jarrpa) [inline pkg/controller/defaults/placements.go:95]:
```suggestion
func getAppWeightedPodAffinityTerm(weight int, selector string) corev1.WeightedPodAffinityTerm {
```

File: pkg/controller/defaults/placements.go [hunk 7]
```diff
@@ -11,219 +11,62 @@ var (
...
-										Operator: metav1.LabelSelectorOpIn,
-										Values:   []string{"rook-ceph-rgw"},
-									},
-								},
-							},
-							TopologyKey: corev1.LabelZoneFailureDomain,
-						},
-					},
-				},
+				RequiredDuringSchedulingIgnoredDuringExecution:  getPodAffinityTerm("rook-ceph-rgw"),
```
Comment (jarrpa) [inline pkg/controller/defaults/placements.go:56]:
```suggestion
				RequiredDuringSchedulingIgnoredDuringExecution:  []corev1.PodAffinityTerm{
					getAppPodAffinityTerm("rook-ceph-rgw"),
				}
```

**Review by jarrpa — CHANGES_REQUESTED** [review]
I agree with a number of Umanga's comments. To start, explain in your commit message (and PR description) what this refactor gains us. Specifically, I see it as reducing the complexity of structures that need to be written by hand.

I also agree that this should probably be more configurable to be more meaningful. As such, I'd like you to take any functions that return a list and only return a single element that would go in that list. I've provided an example in a comment below.

**Comment by rajatsing** [comment]
I considered your changes @jarrpa and I think if I start making changes that you suggested, it will make code more complicated and in turn, the whole point of "refactoring" will be of no use.
I think It's better to close it?.
@umangachapagain wdyt?.

**Comment by jarrpa** [comment]
I completely disagree. Refactoring is more than just simple code. I hate seeing the same exact framework or even exact configuration duplicated in multiple places. While Umanga has a point that these things don't change very frequently (if some would at all), that is no reason to just let the error potential sit around unaddressed. It's all just BZs waiting to come up in QE.

File: pkg/controller/defaults/placements.go (hunk 5 above)
Comment (rajatsing) [reply pkg/controller/defaults/placements.go]: added, please check.

Comment (rajatsing) [reply pkg/controller/defaults/placements.go]: added, please check!.

File: pkg/controller/defaults/placements.go (hunk 6 above)
Comment (rajatsing) [reply pkg/controller/defaults/placements.go]: added, please check!.

File: pkg/controller/defaults/placements.go (hunk 7 above)
Comment (rajatsing) [reply pkg/controller/defaults/placements.go]: added, please check!.

**Comment by rajatsing** [comment]
@jarrpa @umangachapagain 
I have added more specific commit and description. Also, based on the suggestion above, I have modified the code in a way that each struct can be modified based on their own individual's value.

**Review by jarrpa — CHANGES_REQUESTED** [review]
Code looks good, now merge the two commits. :) We don't need to know how you reacted to the review comments in the commit history, just have a clean commit that can be understood (and reverted later if needed).

**Comment by rajatsing** [comment]
> Code looks good, now merge the two commits. :) We don't need to know how you reacted to the review comments in the commit history, just have a clean commit that can be understood (and reverted later if needed).

Absolutely!. When I was reviewing the code myself, I forgot about to review the older commit sadly. But nvm, I'll squash them and form a new commit.

**Comment by rajatsing** [comment]
Commit fixed!.

**Comment by rajatsing** [comment]
I hate it when CI runs on local but fails on remote 👎 .

**Review by jarrpa — CHANGES_REQUESTED** [review]
One last thing I failed to catch earlier: Instead of taking a `string` and setting it as an element of a `[]string{}`, take `...string` and set it directly as the full list.

**Review by jarrpa — APPROVED** [review]
LGTM, thank you!

**Pull request merged**